spec/AuthenticationAdapters.spec.js

```diff
@@ -2140,3 +2152,990 @@ describe('facebook limited auth adapter', () => {
+      authData: {
+        baseAdapter: { id: 'baseAdapter' },
+      },
+    });
+
+    expect(baseAdapter2.validateAuthData).toHaveBeenCalledTimes(1);
+
+    const user3 = new Parse.User();
+    await user3.save({
+      authData: {
+        baseAdapter: { id: 'baseAdapter' },
+        baseAdapter2: { token: true },
+      },
+    });
+
+    expect(baseAdapter2.validateAuthData).toHaveBeenCalledTimes(2);
+  });
+  it('should require additional provider if configured', async () => {
+    await reconfigureServer({
+      auth: { baseAdapter, additionalAdapter },
+    });
+
+    const user = new Parse.User();
+
+    await user.save({
+      authData: {
+        baseAdapter: { id: 'baseAdapter' },
+        additionalAdapter: { token: true },
+      },
+    });
+
+    const user2 = new Parse.User();
+    await expectAsync(
+      user2.save({
+        authData: {
+          baseAdapter: { id: 'baseAdapter' },
+        },
+      })
+    ).toBeRejectedWithError('Missing additional authData additionalAdapter');
+    expect(user2.getSessionToken()).toBeUndefined();
+
+    await user2.save({
+      authData: {
+        baseAdapter: { id: 'baseAdapter' },
+        additionalAdapter: { token: true },
+      },
+    });
+
+    expect(user2.getSessionToken()).toBeDefined();
+  });
+  it('should skip additional provider if used provider is solo', async () => {
+    await reconfigureServer({
+      auth: { soloAdapter, additionalAdapter },
+    });
+
+    const user = new Parse.User();
+
+    await user.save({
+      authData: {
+        soloAdapter: { id: 'soloAdapter' },
+        additionalAdapter: { token: true },
+      },
+    });
+
+    const user2 = new Parse.User();
+    await user2.save({
+      authData: {
+        soloAdapter: { id: 'soloAdapter' },
+      },
+    });
+    expect(user2.getSessionToken()).toBeDefined();
+  });
+  it('should return authData response and save some info on non username login', async () => {
+    spyOn(baseAdapter, 'validateAuthData').and.resolveTo({
+      response: { someData: true },
+    });
+    spyOn(baseAdapter2, 'validateAuthData').and.resolveTo({
+      response: { someData2: true },
+      save: { otherData: true },
+    });
+    await reconfigureServer({
+      auth: { baseAdapter, baseAdapter2 },
+    });
+
+    const user = new Parse.User();
+
+    await user.save({
+      authData: {
+        baseAdapter: { id: 'baseAdapter' },
+        baseAdapter2: { test: true },
+      },
+    });
+
+    expect(user.get('authDataResponse')).toEqual({
+      baseAdapter: { someData: true },
+      baseAdapter2: { someData2: true },
+    });
+
+    const user2 = new Parse.User();
+    user2.id = user.id;
+    await user2.save(
+      {
+        authData: {
+          baseAdapter: { id: 'baseAdapter' },
+          baseAdapter2: { test: true },
+        },
+      },
+      { sessionToken: user.getSessionToken() }
+    );
+
+    expect(user2.get('authDataResponse')).toEqual({ baseAdapter2: { someData2: true } });
```

Comment:
Can you explain this one?

Reply:
Here we check that authDataResponse works also when the user is logged in and send some authData (here baseAdapter2)
` user2.id = user.id;` and `{ sessionToken: user.getSessionToken() }`